cashu/core/b_dhke.py

```diff
@@ -37,43 +57,86 @@
-    If it does not, it tries computing a new point from the hash."""
+    If it does not, iteratively tries to compute a new point from the hash."""
     point = None
     msg_to_hash = message
     while point is None:
+        _hash = hashlib.sha256(msg_to_hash).digest()
         try:
-            _hash = hashlib.sha256(msg_to_hash).digest()
+            # will error if point does not lie on curve
             point = PublicKey(b"\x02" + _hash, raw=True)
         except:
             msg_to_hash = _hash
     return point
 
 
 def step1_alice(
-    secret_msg: str, blinding_factor: bytes = None
+    secret_msg: str, blinding_factor: bytes = b""
 ) -> tuple[PublicKey, PrivateKey]:
     Y: PublicKey = hash_to_curve(secret_msg.encode("utf-8"))
     if blinding_factor:
         r = PrivateKey(privkey=blinding_factor, raw=True)
     else:
         r = PrivateKey()
-    B_: PublicKey = Y + r.pubkey
+    B_: PublicKey = Y + r.pubkey  # type: ignore
     return B_, r
 
 
-def step2_bob(B_: PublicKey, a: PrivateKey) -> PublicKey:
-    C_: PublicKey = B_.mult(a)
-    return C_
+def step2_bob(B_: PublicKey, a: PrivateKey):
+    C_ = B_.mult(a)  # type: ignore
+
+    # produce dleq proof
+    e, s = step2_bob_dleq(B_, a)
+    return C_, e, s
 
 
 def step3_alice(C_: PublicKey, r: PrivateKey, A: PublicKey) -> PublicKey:
-    C: PublicKey = C_ - A.mult(r)
+    C: PublicKey = C_ - A.mult(r)  # type: ignore
     return C
 
 
 def verify(a: PrivateKey, C: PublicKey, secret_msg: str) -> bool:
     Y: PublicKey = hash_to_curve(secret_msg.encode("utf-8"))
-    return C == Y.mult(a)
+    return C == Y.mult(a)  # type: ignore
+
+
+def hash_e(R1: PublicKey, R2: PublicKey, K: PublicKey, C_: PublicKey):
+    _R1 = R1.serialize(compressed=False).hex()
+    _R2 = R2.serialize(compressed=False).hex()
+    _K = K.serialize(compressed=False).hex()
+    _C_ = C_.serialize(compressed=False).hex()
+    e_ = f"{_R1}{_R2}{_K}{_C_}"
+    e = hashlib.sha256(e_.encode("utf-8")).digest()
+    return e
+
+
+def step2_bob_dleq(B_: PublicKey, a: PrivateKey, p_bytes: bytes = b""):
+    if p_bytes:
+        # deterministic p for testing
+        p = PrivateKey(privkey=p_bytes, raw=True)
+    else:
+        # normally, we generate a random p
+        p = PrivateKey()
+
+    R1 = p.pubkey  # R1 = pG
+    assert R1
+    R2 = B_.mult(p)  # R2 = pB_ # type: ignore
+    C_ = B_.mult(a)  # C_ = aB_ # type: ignore
+    K = a.pubkey
+    assert K
+    e = hash_e(R1, R2, K, C_)  # e = hash(R1, R2, K, C_)
+    s = p.tweak_add(a.tweak_mul(e))  # s = p + ek
+    return e, s
+
+
+def alice_verify_dleq(e: bytes, s: bytes, K: PublicKey, B_: bytes, C_: bytes):
+    epk = PrivateKey(e, raw=True)
+    spk = PrivateKey(s, raw=True)
+    bk = PublicKey(B_, raw=True)
+    ck = PublicKey(C_, raw=True)
+    R1 = spk.pubkey - K.mult(epk)  # type: ignore
+    R2 = bk.mult(spk) - ck.mult(epk)  # type: ignore
+    return e == hash_e(R1, R2, K, ck)

```

Comment:
new crypto functions

Reply:
you know, i kinda wonder if `compressed=False` makes much sense in this context. it's just more bytes to hash.
maybe a binary serialization of compressed points would be slightly more optimal?